rclcpp/src/rclcpp/executor.cpp

```diff
@@ -885,6 +885,8 @@ Executor::get_next_executable(AnyExecutable & any_executable, std::chrono::nanos
     // Wait for subscriptions or timers to work on
     wait_for_work(timeout);
     if (!spinning.load()) {
+      // Clear wait result to release ownership of entity
+      wait_result_.reset();
```

Comment:
Actually, this might not be the best place to do this. It is true that if spinning is false, then we no longer need the wait result, but it's possible for the executor to exit without going through this code path. For example, if you exit a spin here:

https://github.com/ros2/rclcpp/blob/5cf00c93f75881a67eb594bb7ec5ee4e07942171/rclcpp/src/rclcpp/executor.cpp#L275-L277 (even if it is an exception)

Or here:

https://github.com/ros2/rclcpp/blob/5cf00c93f75881a67eb594bb7ec5ee4e07942171/rclcpp/src/rclcpp/executor.cpp#L279

Which would result in the same problem you observed in the original issue. Instead I think we should go through the various spin methods and find the right spot to clear the wait result when exiting them, maybe with a `RCPPUTILS_SCOPE_EXIT()` block. The spin functions already clear the results when starting to avoid working on out-of-date information, e.g.:

https://github.com/ros2/rclcpp/blob/5cf00c93f75881a67eb594bb7ec5ee4e07942171/rclcpp/src/rclcpp/executor.cpp#L369-L374

@mjcarroll what do you think?

Reply:
Thanks. These paths need to be considered as well.